irohad/ametsuchi/pointer_cache.hpp

```diff
@@ -0,0 +1,40 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_CONSENSUS_CACHE_H
+#define IROHA_CONSENSUS_CACHE_H
+
+#include <memory>
+
+/**
+ * Stores pointer to some data
+ */
+template <typename DataType,
+          template <typename> class PointerType = std::shared_ptr>
+class PointerCache {
```

Comment:
Is there a need for a base class? Why would we want polymorphism here? Does base class require all of its descendants to be thread safe? If yes maybe use NVI